File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import defaultdict, deque

def one_diff(w1,w2):
    diff = 0
    for idx in range(len(w1)):
        if w1[idx] != w2[idx]:
            diff += 1
    if diff == 1:
        return True
    else:
        return False

def solution(begin, target, words):
    answer = 0
    
    visited = [False] * (len(words))
    q = deque()
    q.append((begin,0))
    while q:
        cur, dist = q.popleft()
        
        if cur == target:
            answer = dist
            break
        
        for i in range(len(words)):
            if not visited[i] and one_diff(words[i], cur):
                q.append((words[i], dist+1))
                visited[i] = True
   
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Node {
    String s;
    int count;
    
    Node(String s, int count) {
        this.s = s;
        this.count = count;
    }
}

class Solution {
    public boolean checkOneDiff(String word1, String word2) {
        int len = word1.length();
        char[] w1 = word1.toCharArray();
        char[] w2 = word2.toCharArray();
        
        int diff = 0;
        for(int i=0;i<len;i++) {
            if (w1[i] != w2[i]) diff++;
        }
        return diff == 1 ? true : false;
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        boolean[] visited = new boolean[words.length];
        Queue<Node> q = new ArrayDeque<>();
        q.add(new Node(begin, 0)); // word,count
        
        while(!q.isEmpty()) {
            Node cur = q.poll();
            if (cur.s.equals(target)) {
                answer = cur.count;
                break;
            }
            
            for(int i=0;i<words.length;i++) {
                String word = words[i];
                if(checkOneDiff(cur.s, word) && !visited[i]) {
                    q.add(new Node(word,cur.count+1));
                    visited[i] = true;
                }
            }
        }
        
        return answer;
    }
}